### PR TITLE
Add UV support to venv operators

### PR DESCRIPTION
**Pull request by jscheffl**
Follow-up on #43553 and #43568

Adds support for UV for the VirtualEnv operators.

Don't know if it is acceptable like this, it is auto-detecting: if UV is installed then it is using UV.
Do we need to make this configurable or is an auto-detection sufficient?

### CODE REVIEW

**Comment by potiuk** [comment]
I'd say:

* add parameter to switch between uv and pip (no auto-detection - too much risk) 
* add unit tests :)

J.

**Comment by jscheffl** [comment]
> I'd say:
> 
>     * add parameter to switch between uv and pip (no auto-detection - too much risk)

With the follow-up question(s):

* What to default?
* Adding a config entry or forcing all DAGs do be adjusted for this param?

File: providers/src/airflow/providers/standard/utils/python_virtualenv.py [hunk 1]
```diff
@@ -88,22 +130,33 @@ def prepare_virtualenv(
     if pip_install_options is None:
         pip_install_options = []
 
+    if requirements is not None and requirements_file_path is not None:
```
Comment (gopidesupavan) [inline providers/src/airflow/providers/standard/utils/python_virtualenv.py:148]:
```suggestion
    if requirements and requirements_file_path:
```

**Review by gopidesupavan — COMMENTED** [review]
small change :)

**Comment by gopidesupavan** [comment]
> > I'd say:
> > ```
> > * add parameter to switch between uv and pip (no auto-detection - too much risk)
> > ```
> 
> With the follow-up question(s):
> 
> * What to default?
> * Adding a config entry or forcing all DAGs do be adjusted for this param?
> 
> Proposal:
> 
> * Option "AUTO", "PIP", "UV" as Enum
>   
>   * AUTO=Like now, checks if UV is installed, if not uses pip
>   * The other options: How the user likes it
> * Default is configurable and uses "AUTO" as default as provider config property
> 
> Alternatively we can also make it like attept UV and if it fails attempt pip as fallback?

I feel like default to UV is better, as now we are completely moving towards UV, so if user want to switch they can update?

File: providers/src/airflow/providers/standard/utils/python_virtualenv.py (hunk 1 above)
Comment (jscheffl) [reply providers/src/airflow/providers/standard/utils/python_virtualenv.py:148]: I _think_ there was a reason behind this and the lines just have been moved in order. There I'd propose to keep it. Even if the proposed shortage looks mot Pythonic there could be cases where Non-Null values evaluate to False.

I just moved the line up to fail fast - else in previous code it was creating the venv base and then raising an exception. Not if fails before venv is started to be created.

The line is originating from long long time ago from https://github.com/apache/airflow/pull/17349

File: airflow/example_dags/example_branch_operator_decorator.py [hunk 2]
```diff
@@ -112,7 +112,7 @@ def some_ext_py_task():
     #       Run the example a second time and see that it re-uses it and is faster.
     VENV_CACHE_PATH = tempfile.gettempdir()
 
-    @task.branch_virtualenv(requirements=["numpy~=1.24.4"], venv_cache_path=VENV_CACHE_PATH)
+    @task.branch_virtualenv(requirements=["numpy~=1.26.0"], venv_cache_path=VENV_CACHE_PATH)
```
Comment (jscheffl) [inline airflow/example_dags/example_branch_operator_decorator.py]: Note: needed to change this as else the example is broken and venv does not work with Python 3.12

Might be worth a separate PR as well to fix it on Airflow 2.10 as well.

**Comment by potiuk** [comment]
> Option "AUTO", "PIP", "UV" as Enum
> AUTO=Like now, checks if UV is installed, if not uses pip

Yes. It's good I think. Needs to be followed up with newsfragment explaining it.

File: providers/src/airflow/providers/standard/utils/python_virtualenv.py (hunk 1 above)
Comment (potiuk) [reply providers/src/airflow/providers/standard/utils/python_virtualenv.py:148]: Yes @jscheffl  is right - requirements migh be "empty" (as result of templating) and this would evaluate to "no requirements" (but `pip install ""` would fail) similarly requirements_file might be templated and empty result would mean "no requirements" rather than trying to do `pip install -r ""` (which would also fail)..

File: airflow/example_dags/example_branch_operator_decorator.py (hunk 2 above)
Comment (potiuk) [reply airflow/example_dags/example_branch_operator_decorator.py]: Yep. Separating is a good idea.

File: providers/src/airflow/providers/standard/utils/python_virtualenv.py (hunk 1 above)
Comment (gopidesupavan) [reply providers/src/airflow/providers/standard/utils/python_virtualenv.py:148]: make sense :)

File: airflow/example_dags/example_branch_operator_decorator.py (hunk 2 above)
Comment (jscheffl) [reply airflow/example_dags/example_branch_operator_decorator.py]: Separated to #43653

File: providers/src/airflow/providers/standard/provider.yaml [hunk 3]
```diff
@@ -62,3 +62,20 @@ hooks:
       - airflow.providers.standard.hooks.filesystem
       - airflow.providers.standard.hooks.package_index
       - airflow.providers.standard.hooks.subprocess
+
+config:
+  standard:
+    description: Options for the standard operators
```
Comment (gopidesupavan) [inline providers/src/airflow/providers/standard/provider.yaml]:
```suggestion
    description: Options for the standard provider operators.
```

File: providers/src/airflow/providers/standard/utils/python_virtualenv.py [hunk 4]
```diff
@@ -88,22 +145,33 @@ def prepare_virtualenv(
...
-        raise ValueError("Either requirements OR requirements_file_path has to be passed, but not both")
-
     pip_cmd = None
     if requirements is not None and len(requirements) != 0:
-        pip_cmd = _generate_pip_install_cmd_from_list(venv_directory, requirements, pip_install_options)
+        if _use_uv():
+            pip_cmd = _generate_uv_install_cmd_from_list(venv_directory, requirements, pip_install_options)
+        else:
+            pip_cmd = _generate_pip_install_cmd_from_list(venv_directory, requirements, pip_install_options)
     if requirements_file_path is not None and requirements_file_path:
```
Comment (gopidesupavan) [inline providers/src/airflow/providers/standard/utils/python_virtualenv.py:166]:
```suggestion
    else:
```
Line: 149 already checks if both requirement/requirement_file_path provided it raise exception, so in this case here else block is fine :) ?

File: providers/src/airflow/providers/standard/utils/python_virtualenv.py [hunk 5]
```diff
@@ -20,15 +20,51 @@
...
 import jinja2
 from jinja2 import select_autoescape
 
+from airflow.configuration import conf
 from airflow.utils.process_utils import execute_in_subprocess
 
 
+def _is_uv_installed() -> bool:
+    """
+    Check if the uv tool is installed via checking if it is on the path or installed as package.
```
Comment (gopidesupavan) [inline providers/src/airflow/providers/standard/utils/python_virtualenv.py]:
```suggestion
    Verify whether the uv tool is installed by checking if it's included in the system PATH or installed as a package.
```

**Review by gopidesupavan — APPROVED** [review]

**Comment by gopidesupavan** [comment]
LGTM :)

File: providers/src/airflow/providers/standard/utils/python_virtualenv.py (hunk 4 above)
Comment (jscheffl) [reply providers/src/airflow/providers/standard/utils/python_virtualenv.py:166]: I also scratched my head regarding this... but the logic can be... use a venv without requirements file and without specified requirements... will install a "plain vanilla" runtime w/o any extra packages. There are even pytests for this. Seems to be a valid requirement having a "minimal with no sugar" venv to execute something within.

Comment (gopidesupavan) [reply providers/src/airflow/providers/standard/utils/python_virtualenv.py:166]: Oh got it :) , thanks jens for explaining.

Comment (jscheffl) [reply providers/src/airflow/providers/standard/utils/python_virtualenv.py:166]: I also learned this while adding the feature :-D Something new every day :-D

Comment (gopidesupavan) [reply providers/src/airflow/providers/standard/utils/python_virtualenv.py:166]: Do you think it would be worthwhile to add comments explaining why those conditions are set up that way? Otherwise, someone might come along later, think they're redundant, and try to remove them?

Comment (jscheffl) [reply providers/src/airflow/providers/standard/utils/python_virtualenv.py:166]: If somebody cleans this up... pytests will fail.

**Comment by set5think** [comment]
@jscheffl WOW THANKS FOR TAKING THIS TO THE FINISH LINE!!!  It's amazing to see how simple it was to add what I think is an extremely valuable contribution!!!  GOOD JOB! :heart:

**Comment by Lee-W** [comment]
just saw it on slack. love this one!
